Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/PredictorCorrector.java

```diff
@@ -15,51 +15,137 @@
 
 package com.amazon.randomcutforest.parkservices;
 
-import static com.amazon.randomcutforest.CommonUtils.checkArgument;
-import static com.amazon.randomcutforest.parkservices.preprocessor.Preprocessor.DEFAULT_NORMALIZATION_PRECISION;
-
-import java.util.Arrays;
-
 import com.amazon.randomcutforest.RandomCutForest;
 import com.amazon.randomcutforest.config.ForestMode;
+import com.amazon.randomcutforest.config.ScoringStrategy;
 import com.amazon.randomcutforest.config.TransformMethod;
+import com.amazon.randomcutforest.parkservices.statistics.Deviation;
 import com.amazon.randomcutforest.parkservices.threshold.BasicThresholder;
 import com.amazon.randomcutforest.returntypes.DiVector;
 import com.amazon.randomcutforest.util.Weighted;
 
+import java.util.Arrays;
+import java.util.Random;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static com.amazon.randomcutforest.CommonUtils.toDoubleArray;
+import static com.amazon.randomcutforest.CommonUtils.toFloatArray;
+import static com.amazon.randomcutforest.parkservices.preprocessor.Preprocessor.DEFAULT_NORMALIZATION_PRECISION;
+import static java.lang.Math.min;
+
 /**
  * This class provides a combined RCF and thresholder, both of which operate in
  * a streaming manner and respect the arrow of time.
  */
 public class PredictorCorrector {
-
     private static double DEFAULT_DIFFERENTIAL_FACTOR = 0.3;
 
     public static int DEFAULT_NUMBER_OF_MAX_ATTRIBUTORS = 5;
 
+    public static double DEFAULT_MULTI_MODE_SAMPLING_RATE = 0.1;
+
+    protected static int NUMBER_OF_MODES = 2;
+
+    // the following vectors enable suppression of anomalies
+    // the first pair correspond to additive differences
+    // the second pair correspond to multiplicative differences
+    // multiplicative differences are not meaningful for differenced operations
```

Comment:
Yes, we will still output the expected values. But those expected values would depend on the previous input, which is known to the PreProcessor, but PredictorCorrector does not know the previous input. The comment here was for the following scenario: Suppose we added 100 to every number --> in differencing nothing changed! So if we test for significance (say 5%) based on 100 then we could be very far off -- because the original numbers can be between 0 and 1 (and so entire data is between 100 and 101). Now, users can correct for that -- but it seemed that this could create more issues

Reply:
Good q's. Let's start with 3. I meant difference context to be "change from one data point to the next in a difference (normalized/otherwise) transform". The discrepancy between actual and expected values are important; but it is hard to interpret that when it is "actual value of the difference and expected value of the difference". Difference transforms are powerful -- but domain dependent.  

Re 2: you are correct. We only run this code when RCF flags an anomaly (and we can compute an expected value). This is why the deviation[] are unused at the moment. Some ideas we can try in the future: randomly sample and evaluate forecast even for points which are not flagged. That is why the randomSeed and nextRandom() are introduced. But it'll likely requite a lot of tuning; I wanted to add those to the state classes now so that we do not have to change state classes/mapper needlessly.

Re 1, the transformations correspond to (i) shift and then (ii) scale; so if there is no differencing involved then an additive difference corresponds to  an additive difference after the shift. The shift varies, but significantly slowly, so the dependence on previous values is very weak.